src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

```diff
@@ -238,7 +238,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <IlcArg Condition="$(IlcDumpIL) == 'true'" Include="--ildump:$(NativeIntermediateOutputPath)%(ManagedBinary.Filename).il" />
       <IlcArg Condition="$(NoWarn) != ''" Include='--nowarn:"$([MSBuild]::Escape($(NoWarn)))"' />
       <IlcArg Condition="$(TrimmerSingleWarn) == 'true'" Include="--singlewarn" />
-      <IlcArg Condition="$(IlcTrimMetadata) == 'true'" Include="--reflectedonly" />
+      <IlcArg Condition="$(IlcTrimMetadata) == 'false'" Include="--reflectableartifacts" />
```

Comment:
That flag name 😅

Reply:
Accepting proposals for a better name. Sometimes I think we should have used hyphens to separate words in command line arguments, but I'm not sure that's the only thing missing here.

Btw, @davidfowl you might be happy to hear that the size of the default `dotnet new webapi` build with all default settings and `<PublishAot>true</PublishAot>` added goes down by 40% after this change.

But it's [broken](https://twitter.com/MStrehovsky/status/1532652675879534593) both before and after and I wasn't in a good enough mental state to debug this yet. With DI the problem and the manifestation of the problem is always _very far_ from each other, which is why I love DI so much.

Reply:
Yea, I'm not in love with the naming but I think the MSBuild property (if there is one) is more important so I wont bikeshed here.

> Btw, @davidfowl you might be happy to hear that the size of the default dotnet new webapi build with all default settings and <PublishAot>true</PublishAot> added goes down by 40% after this change.

Music to my ears 😄 

> But it's [broken](https://twitter.com/MStrehovsky/status/1532652675879534593) both before and after and I wasn't in a good enough mental state to debug this yet. With DI the problem and the manifestation of the problem is always very far from each other, which is why I love DI so much.

This isn't related to DI, it's because of how application parts work in MVC. I can ping you an explain how it works and why that exception is happenining.

Reply:
> This isn't related to DI, it's because of how application parts work in MVC. I can ping you an explain how it works and why that exception is happenining.

I'll probably take you up on that when I sit down to investigate. My current theory is that ASP.NET relies on the fact that by default we set `<TrimmerDefaultAction>copy</TrimmerDefaultAction>` (keep assemblies that are not manifested as trimmable even if nothing references them) and NativeAOT currently has an incompatiblity where it treats `copy` as `copyused` (i.e. we still delete assemblies that are not referenced from anywhere). Can't confirm it because attempting to get a repro without NativeAOT hits an illink bug (#2820).

On the runtime team we would eventually like to see `TrimmerDefaultAction` to not be set to `copy` by default (maybe at least `copyused`, or even `link` by default, so that we can actually trim stuff). If my theory is true, it might be nice to do something on the ASP.NET side so that this pattern isn't broken when we start trimming more.

Reply:
There's a build time task that looks at all referenced assemblies (transitively) that reference Microsoft.AspNetCore.Mvc and we store an assembly attribute in the project's assembly with the name of the referenced assembly (these can also be added dynamically but I'll ignore that for now, yay plugins). At startup we get these attributes, load the assembly and start looking for "application parts". These parts can contain, controllers, compiled razor views, etc but for WebAPI the 90% case is looking for controller types in these assemblies at runtime.

This is an example of what gets burned into the default project for web api project:

<img width="430" alt="image" src="https://user-images.githubusercontent.com/95136/172080988-e794334b-276b-4f4b-b465-11a756f6183b.png">

This pattern is pretty incompatible with trimming at the moment.

Reply:
> Accepting proposals for a better name.

maybe `—keepadditionalmetadata`? New name even more confusing then previous one.

Reply:
> > Accepting proposals for a better name.
> 
> maybe `—keepadditionalmetadata`? New name even more confusing then previous one.

What this switch does is that it generates reflection metadata for any native artifacts (methods, fields, etc.) that were somehow generated (e.g. there was a call to a method, etc.). I would like the switch name to somehow capture that.

Reply:
Maybe `keepreflectedartifacts` to indicate that you keep them. Also do you have other word instead of `artifact` that’s what confuse me the most. I cannot get context for that word from reading the flag name. Maybe `fullreflectionartifacts`? Just throwing ideas around.

Reply:
how about `--reflectiondata=all(generated/compiled)|only(statically)Used|none`

Reply:
I like `--reflectiondata=all(generated/compiled)|only(statically)Used|none` since that that's much easier to understand what's going on, and should play nice with `more usable reflection-free mode` https://github.com/dotnet/runtime/issues/67193

Reply:
@davidfowl So if we teach trimmer that when it sees `ApplicationPartAttribute` it should keep the assembly it points to, it should work? Does it mean it needs to keep the entire assembly, or just parts from it?

Side note: since it already does sort of "source gen" (injects attributes), it might as well do the whole source gen and avoid reflection at runtime. But I don't know enough about this to tell if it's even possible.

Reply:
> how about `--reflectiondata=all(generated/compiled)|only(statically)Used|none`

We have a winner! Thanks!